internal/hcsoci/credentials.go

```diff
@@ -0,0 +1,115 @@
+// +build windows
+
+package hcsoci
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	"github.com/Microsoft/hcsshim/internal/log"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	"github.com/sirupsen/logrus"
+)
+
+// CCGState stores the id used when creating a ccg instance. Used when
+// closing a container to be able to release the instance.
+type CCGState struct {
+	// ID of container that instance belongs to.
+	id string
+}
+
+// Release calls into hcs to remove the ccg instance. These do not get cleaned up automatically
+// they MUST be explicitly removed with a call to ModifyServiceSettings. The instances will persist
+// unless vmcompute.exe exits or they are removed manually as done here.
+func (instance *CCGState) Release(ctx context.Context) error {
+	if err := removeCredentialGuard(ctx, instance.id); err != nil {
+		log.G(ctx).WithFields(logrus.Fields{
+			logrus.ErrorKey: err,
+			"containerID":   instance.id,
+		}).Error("failed to remove container credential guard instance")
+		return err
+	}
+	return nil
+}
+
+// CreateCredentialGuard creates a container credential guard instance and returns the state object to be placed in a v2 container doc.
+func CreateCredentialGuard(ctx context.Context, id, credSpec string, hypervisorIsolated bool) (*hcsschema.ContainerCredentialGuardState, error) {
+	log.G(ctx).WithField("containerID", id).Debug("creating container credential guard instance")
+	// V2 schema ccg setup a little different as its expected to be passed
+	// through all the way to the gcs. Can no longer be enabled just through
+	// a single property. The flow is as follows
+	// ------------------------------------------------------------------------
+	// 1. Call HcsModifyServiceSettings with a ModificationRequest set with a
+	// ContainerCredentialGuardAddInstanceRequest. This is where the cred spec
+	// gets passed in. Transport either "LRPC" (Argon) or "HvSocket" (Xenon).
+	// 2. Query the instance with a call to HcsGetServiceProperties with the
+	// PropertyType "ContainerCredentialGuard". This will return all instances
+	// 3. Parse for the id of our container to find which one correlates to the
+	// container we're building the doc for, then add to the V2 doc.
+	// 4. If xenon container the hvsocketconfig will need to be in the UVMs V2 schema
+	// HcsComputeSystem document before being created/sent to HCS. It must be in the doc
+	// at creation time as we do not support hot adding hvsocket service table entries.
+	// This is currently a blocker for adding support for hyper-v gmsa.
+	transport := "LRPC"
+	if hypervisorIsolated {
+		// TODO(Dcantah) Set transport to HvSocket here when this is supported
+		return nil, fmt.Errorf("v2 hyper-v isolated containers do not support gmsa")
```

Comment:
I think I left another comment about this error string, did you happen to see it?

Reply:
I might be misremembering if I left it yet or not. :)

Basically two things:
- We should prefer to say "hypervisor isolated" rather than "hyper-v isolated".
- "v2" is pretty vague about what it actually means. You and I know it means "container created the v2 HCS schema", but it could refer to something else instead. I would be more clear in this wording and say "hypervisor isolated containers with v2 HCS schema do not support GMSA"